_posts/gRPC-Go performance improvements.md

```diff
@@ -0,0 +1,148 @@
+
+During our implementation we experimented with different parameters like instead of putting our window size to twice the bdp sample we put it to 4/3 times the bdp sample. Note although this is more accurate (since the noted bdp was smaller than or equal to 1.5 the real bandwidth and ideally we want to be twice the real bandwidth so multiplying the sample with 4/3 should put us there), it's not the best option to choose since the growth is slow and it takes some tens of RPCs to ramp up.
+
+Given that we're always bound by the flow control of TCP which for most cases is upper bounded at 4MB, we bound the growth of our window sizes by the same number: 4MB.
+
+BDP estimation and dynamically adjusting window sizes is turned-on by default and can be turned off by setting values manually for connection and/or stream window sizes.
+
+
+## Near-future efforts
+
+
+### Flush syscalls
+
+We noticed a bug in our transport layer which causes us to make a flush syscall for every data frame we write, even if the same goroutine has more data to send. We can batch a lot of these writes to use only one flush. This in fact will not be a big change to the code itself. The performance gain noted by this change is: // numbers
+
+In our efforts to get rid of unnecessary flushes we recently combined the headers and data write for unary and server streaming RPCs to one flush on the client-side. // PR
+
+Another related idea proposed by one of our users (yay community!) was to combine a server response to a unary RPC into one flush. We are currently looking into that as well.
+
+
+### Memory allocation
+
+For every data frame read from the wire a new memory allocation takes place. The same holds true at the gRPC layer for every new message for decompressing and decoding. These allocations result in excessive garbage collection cycles, which are expensive, and can be reduced if we reuse memory buffers. We are working on an idea to create a cache of these allocated buffers and reuse them when possible. We realize that these buffers are different sizes and using the maximum size for all of them will result in wasting a lot of memory. Therefore, we intend to allocate and cache buffers of fixed sizes(tentatively powers of two). So now when reading x bytes from the wire we can find the nearest power of 2 greater than x and reuse a buffer from our cache if available or allocate a new one if need be. We will be using golang sync Pools so we don't have to worry about garbage collection. However, we will need to run sufficient tests before committing to this.
```

Comment:
are expensive. Reusing memory buffers can reduce this GC pressure, and we are prototyping approaches to do so. As requests need buffers of differing sizes, one approach would be to maintain individual memory pools of fixed sizes (powers of two)..